From e841f19dc45b2f92139cff996be2076408fb1c4a Mon Sep 17 00:00:00 2001
From: qizhong cheng <qizhong.cheng@mediatek.com>
Date: Mon, 15 Aug 2022 17:56:11 +0800
Subject: [PATCH 95/95] ANDROID: fix add vendor hooks for unusual abort cases

Add hooks to check the do_serror can be recover or not.
When PCIe encounters completion timeout, PCIe hardware will
reply slave error to bus, which will causes the system to panic.

Bug: 247846368

Signed-off-by: qizhong cheng <qizhong.cheng@mediatek.com>
Change-Id: Iaef6c9a6ac07e552939c0f8e386213cb930cfb6e
---
 arch/arm64/kernel/traps.c      | 7 +++++++
 drivers/android/vendor_hooks.c | 1 +
 include/trace/hooks/traps.h    | 4 ++++
 3 files changed, 12 insertions(+)

diff --git a/arch/arm64/kernel/traps.c b/arch/arm64/kernel/traps.c
index 61b1a0fdcfcb..63570011681b 100644
--- a/arch/arm64/kernel/traps.c
+++ b/arch/arm64/kernel/traps.c
@@ -977,6 +977,13 @@ bool arm64_is_fatal_ras_serror(struct pt_regs *regs, unsigned long esr)
 
 void do_serror(struct pt_regs *regs, unsigned long esr)
 {
+	int ret = 0;
+
+	/* Add vendor hooks for unusual abort cases */
+	trace_android_rvh_do_serror(regs, esr, &ret);
+	if (ret != 0)
+		return;
+
 	/* non-RAS errors are not containable */
 	if (!arm64_is_ras_serror(esr) || arm64_is_fatal_ras_serror(regs, esr))
 		arm64_serror_panic(regs, esr);
diff --git a/drivers/android/vendor_hooks.c b/drivers/android/vendor_hooks.c
index c894bfb7e4a8..5007228c079e 100644
--- a/drivers/android/vendor_hooks.c
+++ b/drivers/android/vendor_hooks.c
@@ -210,6 +210,7 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_do_undefinstr);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_do_ptrauth_fault);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_panic_unhandled);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_arm64_serror_panic);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_do_serror);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_sha256);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_aes_expandkey);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_aes_encrypt);
diff --git a/include/trace/hooks/traps.h b/include/trace/hooks/traps.h
index 4313a014d3ee..4a619716bcdf 100644
--- a/include/trace/hooks/traps.h
+++ b/include/trace/hooks/traps.h
@@ -27,6 +27,10 @@ DECLARE_RESTRICTED_HOOK(android_rvh_arm64_serror_panic,
 	TP_PROTO(struct pt_regs *regs, unsigned int esr),
 	TP_ARGS(regs, esr), 1);
 
+DECLARE_RESTRICTED_HOOK(android_rvh_do_serror,
+	TP_PROTO(struct pt_regs *regs, unsigned int esr, int *ret),
+	TP_ARGS(regs, esr, ret), 1);
+
 #endif /* _TRACE_HOOK_TRAPS_H */
 /* This part must be outside protection */
 #include <trace/define_trace.h>
-- 
2.25.1

